File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = 0
    dun = len(dungeons)
    for perm in permutations(dungeons, dun):
        cnt = 0
        hp = k #각 순열마다 k초기화
        
        for pm in perm:
            if hp >= pm[0]:
                hp -= pm[1]
                cnt += 1
            if cnt > answer:
                answer = cnt
                
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int max = 0;
    static boolean[] visited;
    
    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(k, 0, dungeons);
        return max;
    }
    
    public static void dfs(int k, int cnt, int[][] dungeons) {
        max = Math.max(max, cnt);
        
        for (int i = 0; i < dungeons.length; i++) {
            int require = dungeons[i][0];
            int cost = dungeons[i][1];
            
            if (!visited[i] && k >= require) {
                visited[i] = true;
                dfs(k - cost, cnt + 1, dungeons);
                visited[i] = false;
            }
        }
    }
}